_posts/2016-12-12-Reactive-Programming-Spring-Reactor.md

```diff
@@ -0,0 +1,527 @@
+
+As in, by being more concurrent you can save a lot of money when using cloud and microservices.
+
+<a name ="contract" />
+
+# Contract
+
+Reactive is non-blocking and messages will never overflow the queue, see for the standard definition [http://www.reactive-streams.org/](http://www.reactive-streams.org/).
+
+>Created by Pivotal, Typesafe, Netflix, Oracle, Red Hat and others.
+
+The scope of Reactive Streams is to find a minimal set of interfaces, methods and protocols that will describe the necessary operations and entities to achieve the goal—asynchronous streams of data with non-blocking [back-pressure](http://www.reactivemanifesto.org/glossary#Back-Pressure).
+With back-pressure, a consumer which can not handle the load of events sends towards it, can communicate this towards the upstream components so these can reduce the load.
+Without back-pressure the consumer would either fail catastrophically or drop events.
+
+<p style="text-align: center;">
+  <img style="max-width: 640px;"  alt="reactive contract" src="/img/reactive/reactive-contract.png">
+</p>
+
+This contract defines to send data `0 .. N`.
+[Publisher](http://www.reactive-streams.org/reactive-streams-1.0.0-javadoc/org/reactivestreams/Publisher.html) is an interface with a `subscribe()` method.
+[Subscriber](http://www.reactive-streams.org/reactive-streams-1.0.0-javadoc/org/reactivestreams/Subscriber.html) has 4 callback methods:
+`onSubscribe()`, `onNext()` (which can be called 0 to N times), `onComplete()` and `onError()`.
+The last two signals (complete and error) are terminal states, no further signals may occur and the subscriber's subscription is considered cancelled.
+
+What is important is the reverse flow and the back-pressure.
+After subscribing, the subscriber gets a subscription which is a kind of 1 on 1 relationship between the subscriber and the publisher with 2 methods: `request` and `cancel`.
+
+* Request: this is the more important one, with this method the subscriber will ask the publisher to send x messages (and not more), a so called 'pull'.
+* Cancel: the subscription is being cancelled.
+
+Spring Reactor focuses on the publisher side of the reactive streaming, as this is the hardest to implement and to get right.
+It provides you with the tools to implement publishers in a back-pressure way.
+
+The publisher is a provider of a potentially unbounded number of sequenced elements, publishing them according to the demand received from its Subscriber(s).
+
+The Reactive Streams specification has been adopted for [java 9](http://gee.cs.oswego.edu/dl/jsr166/dist/docs/java/util/concurrent/Flow.html).
+
+# DIY Reactive Streams
+
+Implementing a Reactive Stream framework yourself is very hard to do, for Stephane Maldini this is the 4th or 5th attempt. 
+For Davik Karnok, the tech lead of RxJava, it is attempt 7 or 8.
+The main difficulty is to make it side effect free.
+
+For example:
+
+```
```

Comment:
For code samples in Java I would use the highlight syntax (examples see the Lagom blogpost). Everything looks better with colours! :-)

For example:
{% highlight java %}
public interface HelloService extends Service {
  ServiceCall<String, NotUsed, String> hello();
  
  ServiceCall<String, GreetingMessage, String> useGreeting();

  @Override
  default Descriptor descriptor() {
    return named("helloservice").with(
        restCall(Method.GET,  "/api/hello/:id",       hello()),
        restCall(Method.POST, "/api/hello/:id",       useGreeting())
      ).withAutoAcl(true);
  }
}
{% endhighlight %}